### PR TITLE
Improvements 

### PR DESCRIPTION
**Pull request by DimitrijeDimitrijevic**

### CODE REVIEW

File: lib/excalt/helpers/request.ex [hunk 1]
```diff
@@ -1,14 +1,22 @@
 defmodule Excalt.Helpers.Request do
+  @moduledoc """
+  Helper functions for parsing data from request/response
+  """
   @doc """
   Extract a specific field from the header of the request.
   """
   def extract_from_header(header, field) do
```
Comment (swerter) [inline lib/excalt/helpers/request.ex:8]: Could you maybe write a test case to ensure it works properly?

File: lib/excalt/vcard/addressbook.ex [hunk 2]
```diff
@@ -3,29 +3,41 @@ defmodule Excalt.Vcard.Addressbook do
   Fetching addressbooks from the carddav server
   """
 
-  # https://www.rfc-editor.org/rfc/rfc6352#section-6.2 ->
-  # According to this, the client applications should support
-  # 3.0, 4.0 versions of vcards.
-  # Content types can be "text/vcard"(3.0 and 4.0) or "application/vcard+json"(4.0)
   defstruct name: nil, url: nil, description: nil, content_types: [], versions: []
 
+  @type t :: %__MODULE__{
```
Comment (swerter) [inline lib/excalt/vcard/addressbook.ex:8]: Thanks, appreciated :)

File: lib/excalt/vcard/contact.ex [hunk 3]
```diff
@@ -1,83 +1,196 @@
...
     case Finch.request(request, ExcaltFinch) do
-      {:ok, %Finch.Response{status: 201, body: body}} ->
-        {:ok, body}
-
-      _ ->
-        :error
+      {:ok, %Finch.Response{status: 201, body: _body, headers: headers}} ->
+        etag = Excalt.Helpers.Request.extract_from_header(headers, "etag")
+        if etag == "" do
+          {etag, contact} = get_changed_contact_raw(authentication, req_url)
```
Comment (swerter) [inline lib/excalt/vcard/contact.ex]: I would rather return an error instead of fetching the old contact. Otherwise you will loose the creation data.

Comment (swerter) [reply lib/excalt/vcard/contact.ex]: If you find that this is how it should behave, please add a comment with the reference.

File: lib/excalt/vcard/contact.ex [hunk 4]
```diff
@@ -1,83 +1,196 @@
...
+          password :: String.t(),
+          addressbook_name :: String.t(),
+          vcf_text :: String.t()
+        ) :: {:ok, etag :: String.t()} | {:error, any()}
   def create(server_url, username, password, addressbook_name, vcf_text) do
-    authentication = Excalt.Vcard.Addressbook.build_authentication_header(username, password)
+    authentication = Excalt.Vcard.Helpers.build_authentication_header(username, password)
     req_url = Excalt.Vcard.UrlHelper.build_url(server_url, username, addressbook_name)
     req_body = vcf_text
     contact_uuid = Elixir.UUID.uuid4()
```
Comment (swerter) [inline lib/excalt/vcard/contact.ex]: Ideally, the uuid should be extracted from the vcard directly and only be newly created if it does not exist there.

File: lib/excalt/vcard/req_builder.ex [hunk 5]
```diff
@@ -48,7 +48,7 @@ defmodule Excalt.Vcard.ReqBuilder do
   end
 
   @doc """
-  get a single contantact by url
+  get a single contantact by providing url from the list of contacts from the addressbook
```
Comment (swerter) [inline lib/excalt/vcard/req_builder.ex]: Small typo: contantact should be contact.

File: lib/excalt/vcard/url_helper.ex [hunk 6]
```diff
@@ -1,14 +1,16 @@
 defmodule Excalt.Vcard.UrlHelper do
   @moduledoc """
-  Building url for carrdav requests.
+  Building urls for carrdav requests.
```
Comment (swerter) [inline lib/excalt/vcard/url_helper.ex]: Small typo: carrdav should be carddav.

File: lib/excalt/vcard/req_builder.ex [hunk 7]
```diff
@@ -68,7 +68,60 @@ defmodule Excalt.Vcard.ReqBuilder do
...
+  end
+
+  @doc """
+  Getting only etags from the server and comparing to one we
+  already have can save us a ton of bandwith. The server should return the
+  response off all etags and urls from the addressbook, so we can compare which
+  of them are created/updated/deleted by other clients, and we can update our
+  local addressbook with newest changes.
+  """
+  def get_etags() do
```
Comment (swerter) [inline lib/excalt/vcard/req_builder.ex:100]: please add another function to fetch the etag of on individual vcard.

**Review by swerter — CHANGES_REQUESTED** [review]
Thanks!

File: lib/excalt/vcard/contact.ex (hunk 3 above)
Comment (DimitrijeDimitrijevic) [reply lib/excalt/vcard/contact.ex]: - This is in fact not way how this library should behave. 
- Responsibility of getting the contact again should be responsibility of the client. 
- Actually it is not an error if the etag is not returned. Returning etag is server responsibility, and on the client is to determine what to do if the etag is missing. 

I am changing this behaviour.

File: lib/excalt/vcard/req_builder.ex (hunk 7 above)
Comment (DimitrijeDimitrijevic) [reply lib/excalt/vcard/req_builder.ex:100]: - After a thinking for a while, there is no need to make another xml request to fetch individual contact with etag. The simple get request is enough for that purpose. 
- Our sabre dav has no plugin for handling get request, therefore cannot handle get requests.

File: lib/excalt/vcard/contact.ex (hunk 4 above)
Comment (DimitrijeDimitrijevic) [reply lib/excalt/vcard/contact.ex]: Yes, it will be in the function head to pass, since extraction and actions will be responsibility of the client.

**Pull request merged**